File: ontrack-repository/src/main/java/net/nemerosa/ontrack/repository/AccountRepository.java
package net.nemerosa.ontrack.repository;

import net.nemerosa.ontrack.model.Ack;
import net.nemerosa.ontrack.model.security.Account;
import net.nemerosa.ontrack.model.security.AccountGroup;
import net.nemerosa.ontrack.model.security.AuthenticationSource;
import net.nemerosa.ontrack.model.security.AuthenticationSourceProvider;
import net.nemerosa.ontrack.model.structure.ID;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import java.util.function.Predicate;

public interface AccountRepository {

    @Deprecated
    boolean checkPassword(int accountId, Predicate<String> check);

    @Deprecated
    Optional<Account> findUserByNameAndSource(String username, AuthenticationSourceProvider sourceProvider);

    @Nullable
    BuiltinAccount findBuiltinAccount(@NotNull String username);

    Collection<Account> findAll(Function<String, AuthenticationSource> authenticationSourceFunction);

    Account newAccount(Account account);

    void saveAccount(Account account);

    Ack deleteAccount(ID accountId);

    void setPassword(int accountId, String encodedPassword);

    Account getAccount(ID accountId, Function<String, AuthenticationSource> authenticationSourceFunction);

    List<Account> findByNameToken(String token, Function<String, AuthenticationSource> authenticationSourceFunction);

    /**
     * Gets the list of accounts associated with this account group.
     *
     * @param accountGroup                 Account group
     * @param authenticationSourceFunction Access to the authentication sources
     * @return List of accounts
     */
    List<Account> getAccountsForGroup(AccountGroup accountGroup, Function<String, AuthenticationSource> authenticationSourceFunction);
}

File: ontrack-repository/src/main/java/net/nemerosa/ontrack/repository/AccountRepository.kt
package net.nemerosa.ontrack.repository

import net.nemerosa.ontrack.model.Ack
import net.nemerosa.ontrack.model.security.Account
import net.nemerosa.ontrack.model.security.AccountGroup
import net.nemerosa.ontrack.model.security.AuthenticationSource
import net.nemerosa.ontrack.model.structure.ID

interface AccountRepository {

    /**
     * Gets the list of all accounts
     */
    fun findAll(): Collection<Account>

    /**
     * Creates a new account
     */
    fun newAccount(account: Account): Account

    /**
     * Edits an existing account
     */
    fun saveAccount(account: Account)

    /**
     * Deletes an account
     */
    fun deleteAccount(accountId: ID): Ack

    /**
     * Changes the password of an account
     */
    @Deprecated("Will be removed in V5")
    fun setPassword(accountId: Int, encodedPassword: String)

    /**
     * Loads an account by ID
     */
    fun getAccount(accountId: ID): Account

    /**
     * Looks for accounts based on some text.
     */
    fun findByNameToken(token: String): List<Account>

    /**
     * Gets the list of accounts associated with this account group.
     *
     * @param accountGroup                 Account group
     * @return List of accounts
     */
    fun getAccountsForGroup(accountGroup: AccountGroup): List<Account>

    /**
     * Finds an account using its name only.
     */
    fun findAccountByName(username: String): Account?

    /**
     * Checks if an account ID exists
     */
    fun doesAccountIdExist(id: ID): Boolean

    /**
     * Deletes all accounts having the given source.
     *
     * @param source Source to delete accounts from
     */
    @Deprecated("Will be removed in V5")
    fun deleteAccountBySource(source: AuthenticationSource)

    /**
     * Disabled / enables an account
     */
    @Deprecated("Will be removed in V5")
    fun setAccountDisabled(id: ID, disabled: Boolean)

    /**
     * Locks / unlocks an account
     */
    @Deprecated("Will be removed in V5")
    fun setAccountLocked(id: ID, locked: Boolean)

    /**
     * Finds or create the account
     */
    fun findOrCreateAccount(account: Account): Account
}